src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/ListPackage/ListPackageCommandRunner.cs

```diff
@@ -0,0 +1,529 @@
+            Dictionary<InstalledPackageReference, IList<NuGetVersion>> uniquePackages)
+        {
+            var sourceRepository = Repository.CreateSource(providers, packageSource, FeedType.Undefined);
+            var dependencyInfoResource = await sourceRepository.GetResourceAsync<FindPackageByIdResource>(listPackageArgs.CancellationToken);
+
+            var packages = (await dependencyInfoResource.GetAllVersionsAsync(package.Name, new SourceCacheContext(), NullLogger.Instance, listPackageArgs.CancellationToken));
+            
+            var latestVersionAtSource = packages.Where(version => MeetsConstraints(version, package, listPackageArgs))
+            .OrderByDescending(version => version, VersionComparer.Default)
+            .FirstOrDefault();
+
+            var latestVersionsForPackage = uniquePackages.Where(p => p.Key.Name.Equals(package.Name, StringComparison.OrdinalIgnoreCase)).Single().Value;
+            latestVersionsForPackage.Add(latestVersionAtSource);
+
+        }
+
+        /// <summary>
+        /// Given a found version from a source and the current version and the args
+        /// of list package, this function checks if the found version meets the required
+        /// highest-patch, highest-minor or prerelease
+        /// </summary>
+        /// <param name="newVersion">Version from a source</param>
+        /// <param name="package">The required package with its current version</param>
+        /// <param name="listPackageArgs">Used to get the constraints</param>
+        /// <returns>Whether the new version meets the constraints or not</returns>
+        private bool MeetsConstraints(NuGetVersion newVersion, InstalledPackageReference package, ListPackageArgs listPackageArgs)
+        {                
+            var result = !newVersion.IsPrerelease || listPackageArgs.Prerelease || package.ResolvedVersion.IsPrerelease;
+
+            if (listPackageArgs.HighestPatch)
+            {
+                result = newVersion.Minor.Equals(package.ResolvedVersion.Minor) && newVersion.Major.Equals(package.ResolvedVersion.Major) && result;
+            }
+
+            if (listPackageArgs.HighestMinor)
+            {
+                result = newVersion.Major.Equals(package.ResolvedVersion.Major) && result;
+            }
+
+            return result;
+        }
+
+
+        /// <summary>
+        /// A function that prints all the package references of a project
+        /// </summary>
+        /// <param name="packages">A list of framework packages. Check <see cref="FrameworkPackages"/></param>
+        /// <param name="projectName">The project name</param>
+        /// <param name="transitive">Whether include-transitive flag exists or not</param>
+        /// <param name="outdated">Whether outdated flag exists or not</param>
+        /// <param name="autoReferenceFound">An out to return whether autoreference was found</param>
+        private void PrintProjectPackages(IEnumerable<FrameworkPackages> packages,
```

Comment:
i would prefer if the printing logic was refactored into a utility class. the runner can have the core logic and be the driver.